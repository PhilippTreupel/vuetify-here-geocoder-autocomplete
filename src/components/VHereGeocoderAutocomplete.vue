<template>
  <v-autocomplete
    v-model="place"
    v-bind="$attrs"
    :items="places"
    :loading="loading"
    :search-input.sync="query"
    :filter="filter"
    return-object
    item-text="title"
    :prepend-icon="prependIcon"
    v-on="$listeners"
    @click:clear="onClear"
  >
    <template slot="item" slot-scope="data">
      <template v-if="typeof data.item !== 'object'">
        <v-list-item-content>{{ data.item }}</v-list-item-content>
      </template>
      <template v-else>
        <v-list-item-content>
          <slot name="highlight" :highlight="data.item.highlight">
            <v-list-item-title v-html="data.item.highlight" />
          </slot>
        </v-list-item-content>
      </template>
    </template>
  </v-autocomplete>
</template>

<script>
export default {
  name: "VHereGeocoderAutocomplete",
  props: {
    value: {
      type: [Object, String],
      required: false,
      default() {
        return {};
      }
    },
    hereApiKey: {
      type: String,
      required: false,
      default: null
    },
    hereBearerOAuthToken: {
      type: String,
      required: false,
      default: null
    },
    at: {
      type: String,
      required: false,
      default: null
    },
    in: {
      type: String,
      required: false,
      default: null
    },
    limit: {
      type: Number,
      required: false,
      default: null
    },
    types: {
      type: String,
      required: false,
      default: null
    },
    lang: {
      type: String,
      required: false,
      default: navigator.language ? navigator.language : null
    },
    resultType: {
      type: String,
      required: false,
      default: null
    },
    houseNumberType: {
      type: String,
      required: false,
      default: null
    },
    localityType: {
      type: String,
      required: false,
      default: null
    },
    administrativeAreaType: {
      type: String,
      required: false,
      default: null
    },
    debounceTime: {
      type: [Boolean, Number, String],
      default: false
    },
    // Vuetify props
    prependIcon: {
      type: String,
      default: "mdi-map-marker"
    },
    customHighlight: {
      type: Function,
      default: placeItem => {
        if (placeItem.highlights.title != null) {
          let ret = placeItem.title;
          // this is necessary because adding a bold tags displaces the highlight indexes delivered by the here api
          let incrementCounter = 0;
          placeItem.highlights.title.forEach(highlight => {
            ret = [
              ret.slice(0, highlight.start + incrementCounter),
              "<b>",
              ret.slice(
                highlight.start + incrementCounter,
                highlight.end + incrementCounter
              ),
              "</b>",
              ret.slice(highlight.end + incrementCounter)
            ].join("");
            // number of characters added with the bold tag
            incrementCounter += 7;
          });
          return ret;
        } else {
          return placeItem.title;
        }
      }
    }
  },
  data() {
    // The initial value can be a string or an object
    // eslint-disable-next-line no-nested-ternary
    const initialValue = this.value
      ? typeof this.value === "string"
        ? this.value
        : this.value.title
      : null;
    const initialPlace = { title: initialValue };
    return {
      loading: false,
      query: initialValue,
      place: initialPlace,
      places: initialValue ? [initialPlace] : [],
      filter() {
        return true; // filtering is done right after the api request
      }
    };
  },
  computed: {
    searchQuery() {
      let finalSearchQuery =
        "https://autocomplete.search.hereapi.com/v1/autocomplete?q=";
      if (this.query != null) {
        finalSearchQuery += this.query.replace(/\s/g, "+");
      }
      if (this.hereBearerOAuthToken == null) {
        finalSearchQuery += "&apiKey=" + this.hereApiKey;
      }
      if (this.at !== null) {
        finalSearchQuery += "&at=" + this.at;
      }
      if (this.in !== null) {
        finalSearchQuery += "&in=" + this.in;
      }
      if (this.limit !== null) {
        finalSearchQuery += "&limit=" + this.limit;
      }
      if (this.types !== null) {
        finalSearchQuery += "&types=" + this.types;
      }
      if (this.lang !== null) {
        finalSearchQuery += "&lang=" + this.lang;
      }
      return finalSearchQuery;
    },
    queryReady() {
      return (
        this.query !== null && this.query !== undefined && this.query !== ""
      );
    },
    fetchCallOptions() {
      if (this.hereBearerOAuthToken == null) {
        return {};
      } else {
        return {
          headers: { Authorization: "Bearer " + this.hereBearerOAuthToken }
        };
      }
    }
  },
  watch: {
    $props: {
      immediate: true,
      handler() {
        this.validateProps();
      }
    },
    query(val) {
      if (val !== null && val !== undefined && val.length >= 1) {
        this.debouncedSearchPlaces();
      }
    },
    debounceTime: {
      immediate: true,
      handler(val) {
        if (val) {
          if (typeof val === "boolean") {
            this.debouncedSearchPlaces = this.debounce(this.searchPlaces, 250);
          } else if (typeof val === "number" || typeof val === "string") {
            this.debouncedSearchPlaces = this.debounce(
              this.searchPlaces,
              Number(val)
            );
          }
        } else {
          this.debouncedSearchPlaces = this.searchPlaces;
        }
      }
    },
    searchQuery: {
      deep: true,
      handler(newVal, oldVal) {
        const oldQuery = oldVal.split("?q=")[1].split("&")[0];
        const newQuery = newVal.split("?q=")[1].split("&")[0];
        if (newQuery !== "" && newQuery === oldQuery) {
          this.debouncedSearchPlaces(this.setPlace);
        }
      }
    }
  },
  mounted() {
    if (this.queryReady) {
      this.searchPlaces(this.emitInputOnMountedSearch);
    }
  },
  methods: {
    async searchPlaces(callback = () => {}) {
      this.loading = true;
      try {
        if (this.queryReady) {
          let searchResults = await (
            await fetch(this.searchQuery, this.fetchCallOptions)
          ).json();
          searchResults = searchResults.items;
          if (searchResults !== null && searchResults !== undefined) {
            this.loading = false;
            if (this.resultType !== null) {
              searchResults = searchResults.filter(item => {
                return (
                  item.resultType !== null &&
                  item.resultType === this.resultType
                );
              });
            }
            if (this.localityType !== null) {
              searchResults = searchResults.filter(item => {
                return (
                  item.localityType !== null &&
                  item.localityType === this.localityType
                );
              });
            }
            if (this.administrativeAreaType !== null) {
              searchResults = searchResults.filter(item => {
                return (
                  item.administrativeAreaType !== null &&
                  item.administrativeAreaType === this.administrativeAreaType
                );
              });
            }
            if (this.houseNumberType !== null) {
              searchResults = searchResults.filter(item => {
                return (
                  item.houseNumberType !== null &&
                  item.houseNumberType === this.houseNumberType
                );
              });
            }
            this.places = searchResults.map(item => {
              const highlight = this.customHighlight(item);
              delete item.highlights;
              return {
                ...item,
                highlight
              };
            });
            callback(this.places[0]);
          } else {
            this.loading = false;
            callback(this.place);
          }
        } else {
          this.loading = false;
          callback(this.place);
        }
      } catch (error) {
        this.loading = false;
        this.$emit("error", error);
      }
    },
    async debouncedSearchPlaces() {},
    onClear() {
      this.$emit("input", null);
      this.$emit("clear");
    },
    emitInputOnMountedSearch(place) {
      this.setPlace(place);
      this.$emit("input", this.place);
    },
    setPlace(place) {
      this.place = place;
    },
    debounce(callback, wait, immediate) {
      let timeout;
      return () => {
        let context = this,
          args = arguments;
        let later = () => {
          timeout = null;
          if (!immediate) callback.apply(context, args);
        };
        let callNow = immediate && !timeout;
        clearTimeout(timeout);
        timeout = setTimeout(later, wait);
        if (callNow) callback.apply(context, args);
      };
    },
    validateProps() {
      // validate that at least one auth method is defined
      if (this.hereApiKey == null && this.hereBearerOAuthToken == null) {
        console.error(
          "Component error <v-here-geocoder-autocomplete>: One of the Here API authentication properties is required and must be defined. \n" +
            "Either 'hereApiKey' with a valid API key or 'hereBearerOAuthToken' with an unexpired OAuth token that can be obtained via the Here API OAuth interface.  "
        );
      }
    }
  }
};
</script>
