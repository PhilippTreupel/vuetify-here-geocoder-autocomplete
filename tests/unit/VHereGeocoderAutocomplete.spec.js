import { mount, createLocalVue } from "@vue/test-utils";
import VHereGeocoderAutocomplete from "@/components/VHereGeocoderAutocomplete";
import Vuetify from "vuetify";
const flushPromises = require("flush-promises");
const sinon = require("sinon");
describe("VHereGeocoderAutocomplete.vue", () => {
  let localVue;
  let wrapper;
  let vuetify;
  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(vuetify);
  });
  it("renders and match snapshot", () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("is searching results for a given entry", async () => {
    let place = "Berlin";
    const searchPlacesMock = sinon.fake();
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        value: place,
        hereApiKey: "HERE_API_KEY"
      },
      methods: {
        searchPlaces: searchPlacesMock
      }
    });
    expect(wrapper.vm.query).toBe(place);
    await flushPromises();
    expect(searchPlacesMock.called).toBe(true);
  });
  it("sets default data properties as expected", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    expect(wrapper.vm.loading).toBe(false);
    expect(wrapper.vm.query).toBeUndefined();
    expect(wrapper.vm.place.title).toBeUndefined();
    expect(wrapper.vm.places.length).toBe(0);
    expect(wrapper.vm.filter()).toBe(true);
  });
  it("sets default component props as expected", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    expect(Object.keys(wrapper.vm.$props.value).length).toBe(0);
    expect(wrapper.vm.$props.hereApiKey).toEqual("HERE_API_KEY");
    expect(wrapper.vm.$props.at).toBeNull();
    expect(wrapper.vm.$props.in).toBeNull();
    expect(wrapper.vm.$props.limit).toBeNull();
    expect(wrapper.vm.$props.types).toBeNull();
    expect(wrapper.vm.$props.lang).toBe(navigator.language);
    expect(wrapper.vm.$props.resultType).toBeNull();
    expect(wrapper.vm.$props.houseNumberType).toBeNull();
    expect(wrapper.vm.$props.localityType).toBeNull();
    expect(wrapper.vm.$props.administrativeAreaType).toBeNull();
    expect(wrapper.vm.$props.debounceTime).toBe(false);
    expect(wrapper.vm.$props.prependIcon).toEqual("mdi-map-marker");
    expect(
      wrapper.vm.$props.customHighlight({
        title: "TESTTESTTEST",
        highlights: {
          title: [
            { start: 0, end: 4 },
            { start: 8, end: 12 }
          ]
        }
      })
    ).toEqual("<b>TEST</b>TEST<b>TEST</b>");
  });
  it("has working debounce function", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    const testFunction = sinon.fake();
    const debounceTestFunction = wrapper.vm.debounce(testFunction, 200);
    debounceTestFunction();
    debounceTestFunction();
    await new Promise(r => setTimeout(r, 300));
    expect(testFunction.calledOnce).toBe(true);
  });
  it("has working debouncedSearchPlaces function", async () => {
    const searchPlacesMock = sinon.fake();
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      },
      methods: {
        searchPlaces: searchPlacesMock
      }
    });
    await wrapper.vm.debouncedSearchPlaces();
    expect(searchPlacesMock.called).toBe(true);
  });
  it("clears input onClear()", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    wrapper.vm.onClear();
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().input[0][0]).toBeNull();
    expect(wrapper.emitted().clear).toBeTruthy();
  });
  it("has working emitInputOnMountedSearch function", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY"
      }
    });
    const place = "TEST";
    wrapper.vm.emitInputOnMountedSearch(place);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.place).toEqual(place);
    expect(wrapper.emitted().input[0][0]).toEqual(place);
  });
  it("has working searchPlaces function", async () => {
    const resultType = "resultType";
    const houseNumberType = "houseNumberType";
    const localityType = "localityType";
    const administrativeAreaType = "administrativeAreaType";
    const foundItem = {
      resultType,
      houseNumberType,
      localityType,
      administrativeAreaType,
      title: "TEST1",
      highlights: {}
    };
    const error = new Error("TEST ERROR");
    require("jest-fetch-mock").enableMocks();
    fetch.enableMocks();
    fetch.mockResponse(
      JSON.stringify({
        items: [
          foundItem,
          {
            title: "TEST2",
            highlights: {}
          },
          {
            resultType,
            houseNumberType,
            localityType,
            administrativeAreaType: "OTHER VALUE",
            title: "TEST3",
            highlights: {}
          }
        ]
      })
    );
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY",
        resultType,
        houseNumberType,
        localityType,
        administrativeAreaType
      }
    });
    wrapper.vm.query = "TEST";
    await wrapper.vm.searchPlaces(firstItem => {
      expect(firstItem.title).toBe(foundItem.title);
    });
    expect(wrapper.vm.places.length).toBe(1);

    // test else block if searchResults==null
    fetch.mockResponseOnce(JSON.stringify({}));
    await wrapper.vm.searchPlaces(place => {
      expect(place).toEqual(wrapper.vm.place);
    });
    // test else block if queryReady === false
    wrapper.vm.query = null;
    await wrapper.vm.searchPlaces(place => {
      expect(place).toEqual(wrapper.vm.place);
    });
    // test the catch block
    fetch.mockRejectOnce(error);
    wrapper.vm.query = "TEST";
    await wrapper.vm.searchPlaces(() => {});
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().error[0][0]).toBe(error);
  });
  it("has a working watcher on computed prop searchQuery", async () => {
    let place = "Berlin";
    const searchPlacesMock = sinon.fake();
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        value: place,
        hereApiKey: "HERE_API_KEY"
      },
      methods: {
        searchPlaces: searchPlacesMock
      }
    });
    expect(wrapper.vm.query).toBe("Berlin");
    await flushPromises();
    // now the actually test of the edge case in searchQuery watcher
    wrapper.vm.$props.limit = 1;
    await flushPromises();
    expect(searchPlacesMock.calledTwice).toBe(true);
  });
  it("has a working watcher on the prop debounceTime", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        debounceTime: true,
        hereApiKey: "HERE_API_KEY"
      }
    });
    expect(wrapper.vm.$props.debounceTime).toBe(true);
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        debounceTime: 250,
        hereApiKey: "HERE_API_KEY"
      }
    });
    expect(wrapper.vm.$props.debounceTime).toBe(250);
  });
  it("configures the searchQuery correctly depending on the prop values", async () => {
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereApiKey: "HERE_API_KEY",
        at: "AT",
        in: "IN",
        limit: 1,
        types: "TYPES",
        lang: "LANG"
      }
    });
    wrapper.vm.query = "QUERY QUERY";
    expect(wrapper.vm.searchQuery).toBe(
      "https://autocomplete.search.hereapi.com/v1/autocomplete?q=QUERY+QUERY&apiKey=HERE_API_KEY&at=AT&in=IN&limit=1&types=TYPES&lang=LANG"
    );
  });
  it("logs an error if none of the auth props are set", async () => {
    jest.spyOn(console, "error").mockImplementationOnce(() => {});
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {}
    });
    expect(console.error).toHaveBeenCalled();
  });
  it("returns the right options for fetch() if 'hereBearerOAuthToken' is set", async () => {
    const authToken = "OAUTH_TOKEN";
    wrapper = mount(VHereGeocoderAutocomplete, {
      localVue,
      vuetify,
      propsData: {
        hereBearerOAuthToken: authToken
      }
    });

    expect(wrapper.vm.fetchCallOptions.headers.Authorization).toEqual(
      "Bearer " + authToken
    );
  });
});
