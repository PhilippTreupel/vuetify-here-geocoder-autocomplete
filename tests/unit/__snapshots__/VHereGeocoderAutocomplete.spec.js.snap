// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VHereGeocoderAutocomplete.vue renders and match snapshot 1`] = `
<div class="v-input theme--light v-text-field v-select v-autocomplete">
  <div class="v-input__prepend-outer">
    <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-map-marker theme--light"></i></div>
  </div>
  <div class="v-input__control">
    <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-2" class="v-input__slot">
      <div class="v-select__slot"><input id="input-2" type="text" autocomplete="off">
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
        </div><input type="hidden" value="[object Object]">
      </div>
      <div class="v-menu">
        <!---->
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
      </div>
    </div>
  </div>
</div>
`;
